### PR TITLE
Update the Autobahn test suite to v0.8.2

### PR DESCRIPTION
**Pull request by anesabml**
## What do these changes do?

This patch updates the Autobahn test suite to v0.8.2.

## Are there changes in behavior for the user?

No

## Related issue number

#4247

Also, this PR incorporates and resolves #5368 by @derlih.

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: CHANGES/4247.misc
Comment (webknjaz) [inline CHANGES/4247.misc]:
```suggestion
Updated the Autobahn test suite to v0.8.2.
```

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [inline tests/autobahn/Dockerfile.aiohttp]: We could probably use something newer, right?

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [inline tests/autobahn/Dockerfile.aiohttp:17]: Why is this necessary?

File: tests/autobahn/Dockerfile.autobahn
Comment (webknjaz) [inline tests/autobahn/Dockerfile.autobahn]: I don't see this being used, is it necessary?

File: tests/autobahn/client/docker-compose.yml
Comment (webknjaz) [inline tests/autobahn/client/docker-compose.yml:1]: I wonder what would it take to integrate this into pytest

**Comment by anesabml** [comment]
@webknjaz I am having some difficulties trying to integrate the test suit with pytest using [pytest-docker-compose](https://pypi.org/project/pytest-docker-compose/) mainly with testing the server, I have asked a question on [stackoverflow](https://stackoverflow.com/questions/67524522/how-to-integrate-autobahn-websocket-client-test-suite-with-pytest) if you want to take a look. (also I am not sure about my implementation).
My suggestion is to use a tool such as [this](https://github.com/marketplace/actions/docker-compose-action) to run the tests without the need of pytest.

https://stackoverflow.com/questions/67524522/how-to-integrate-autobahn-websocket-client-test-suite-with-pytest

File: CHANGES/4247.misc
Comment (webknjaz) [reply CHANGES/4247.misc]: ^

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [reply tests/autobahn/Dockerfile.aiohttp:17]: ^

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [inline tests/autobahn/Dockerfile.aiohttp]: This doesn't look necessary

**Comment by webknjaz** [comment]
@anesabml did you remove your previous comment? If you can't get this into pytest, we could postpone doing it and you could just try adding separate jobs in the CI maybe calling the same things directly.

**Comment by anesabml** [comment]
@webknjaz integrating with pytest is complicated because pytest-docker-compose usage shines when you have a container that you want to run the test against which is the case when we are testing the client, but for testing the server it's the opposite, which is why I couldn't figure out how to do it yet, which got me thinking if we could drop pytest-docker-compose and use GitHub Actions instead.

**Comment by webknjaz** [comment]
You could always just use subprocess invocations and orchestrate spinning the containers from tests (maybe your own fixtures encapsulating this)

**Comment by webknjaz** [comment]
For now, please address/comment on my comments, even w/o a CI, and let's move any further integrations to a separate PR.

**Comment by anesabml** [comment]
@webknjaz  Should I create an issue first or just open a PR for the integration into CI?

**Comment by webknjaz** [comment]
It doesn't really matter, a PR should be enough.

**Comment by webknjaz** [comment]
Looks like you marked the comments as resolved but didn't actually make any changes related to them.

**Comment by anesabml** [comment]
OK, I will open a PR.

**Comment by anesabml** [comment]
What comments did I miss? I think I only missed the one about the changes file.

**Comment by webknjaz** [comment]
> What comments did I miss? I think I only missed the one about the changes file.

I usually expect that the PR author opens each collapsed comment in the files tab to check them. Because this is exactly what I'd do otherwise and it just wastes time because you'll need to look at them anyway but after I find time to go through this. This essentially increases the global time to merge for no good reason.

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [reply tests/autobahn/Dockerfile.aiohttp:17]: @anesabml this thread doesn't have any response from you.

File: tests/autobahn/client/docker-compose.yml
Comment (webknjaz) [reply tests/autobahn/client/docker-compose.yml:1]: @anesabml add a link to the new PR here, once you have it. This is useful for interlinking related changes.

**Review by webknjaz — COMMENTED** [review]
:point_down:

**Comment by anesabml** [comment]
@webknjaz I want to open the new PR concerning automating autobahn tests and stack it on top of this PR, however, I don't know how to do so because the autobahn branch isn't present in this repository.

**Comment by webknjaz** [comment]
It's in your fork, right?

**Comment by anesabml** [comment]
OK, I will open a PR against my fork and change the base branch after this PR gets merged.

**Comment by webknjaz** [comment]
I think you could finish this PR and I'd merge it first

**Comment by anesabml** [comment]
Is there anything you want me to add to this PR? one more thing, can you please check my responses to the comments above?

**Comment by webknjaz** [comment]
@anesabml at least, there are merge conflicts that need to be addressed to unblock me.

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [reply tests/autobahn/Dockerfile.aiohttp:17]: @anesabml ^

**Comment by anesabml** [comment]
@webknjaz Should we add a function to extract the results from the generated reports, such as showing how many tests passed and failed and some info about the failing tests?

**Comment by webknjaz** [comment]
Maybe. But first go through those comments. I need replies.

**Comment by anesabml** [comment]
@webknjaz I am working on building and pushing a docker image to GitHub container registry instead of building it while running the tests, I have created a workflow but it's failing, so this is currently a WIP.

**Comment by webknjaz** [comment]
Maybe submit that separately then?

**Comment by anesabml** [comment]
@webknjaz This was suggested in the original PR, this why I want to include It in this PR.

**Comment by webknjaz** [comment]
But if the code will rely on the container that is not published, it'll keep failing and it's impossible to test it without the container before merging. So the container building and publishing should be merged first.

**Comment by anesabml** [comment]
Yes, I totally missed that. I'll remove the changes.

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [reply tests/autobahn/Dockerfile.aiohttp:17]: Reminder to reply here.

**Comment by webknjaz** [comment]
Also, looking at the image, it'll end up being useless in the current form because it will always embed the old version of aiohttp while it's required to test the current one. It'd take a separate effort to redesign this, embedding only the runtime and build deps, and maybe making possible to use pre-built wheels instead of compiling aiohttp on install.

File: tests/autobahn/Dockerfile.aiohttp
Comment (anesabml) [reply tests/autobahn/Dockerfile.aiohttp:17]: the `attrs` packaged is used [here](https://github.com/aio-libs/aiohttp/blob/4d58400477331704132cc9427d66699e5815ee25/aiohttp/web_protocol.py#L23) I think this is why the original author used `dataclass` instead

File: tests/autobahn/client/docker-compose.yml
Comment (anesabml) [reply tests/autobahn/client/docker-compose.yml:1]: Here's is the [PR](https://github.com/anesabml/aiohttp/pull/1)

File: tests/autobahn/Dockerfile.aiohttp
Comment (anesabml) [reply tests/autobahn/Dockerfile.aiohttp]: I changed it to python:3.9.5

**Comment by anesabml** [comment]
Yes, I have thought about that. Do you have any idea how to approach this problem?

**Comment by webknjaz** [comment]
I think it should be a separate PR because this one has too many changes and discussions. And regarding how to approach the problem — you'd need an image that only has the runtime deps and maybe build-time ones too. Then, in CI you'd invoke that container with the source dir and dists mounted as a volume + run `pip install` on the wheeldir to install what's pre-built and only then run the testing with that.

File: tests/autobahn/client/docker-compose.yml
Comment (webknjaz) [reply tests/autobahn/client/docker-compose.yml:1]: Oh, I thought you'd send it against upstream.

File: tests/autobahn/Dockerfile.aiohttp
Comment (webknjaz) [reply tests/autobahn/Dockerfile.aiohttp:17]: Alright. So it looks like https://github.com/aio-libs/aiohttp/pull/5284 attempted to wipe any use of attrs across the repo but @asvetlov missed this file. This probably needs to be taken care of: https://github.com/aio-libs/aiohttp/issues/5806.

Please add a code comment next to this line explaining the need.

Comment (webknjaz) [reply tests/autobahn/Dockerfile.aiohttp:17]:
```suggestion
# NOTE: The attrs lib is necessary because it's
# NOTE: still a de-facto runtime dependency of
# NOTE: aiohttp. Although it shouldn't be anymore
# NOTE: after the effort to get rid of it.
# NOTE: It must be removed once made redundant.
# Refs:
# * https://github.com/aio-libs/aiohttp/issues/5806
# * https://github.com/aio-libs/aiohttp/pull/5679/files#r654428147
# * https://github.com/aio-libs/aiohttp/pull/5284
RUN pip install attrs
```

**Review by webknjaz — APPROVED** [review]

**Comment by anesabml** [comment]
Thank you for the suggestion, I will try to work on this when I finish the integration of autobahn tests with pytest.

File: tests/autobahn/client/docker-compose.yml
Comment (anesabml) [reply tests/autobahn/client/docker-compose.yml:1]: Here's the new PR against upstream https://github.com/aio-libs/aiohttp/pull/5809

**Pull request merged**
